### PR TITLE
(Migration) TMI.js -> Twurple

### PR DESCRIPTION
**Pull request by AceLikesGhosts**
Starting the migration to Twurple.

### CODE REVIEW

**Comment by AceLikesGhosts** [comment]
There is an issue with this, which at the moment I'm not going to attempt to fix; On our Discord side if we send a message we never actually bind the Twitch message to the linkedListNode, so we are unable to delete it. Other than that this is a completely working migration. Can someone else please look into it? I believe I might have messed up some logic within our twitch.ts for that.

**Comment by AceLikesGhosts** [comment]
Oop, relooking at my changes and I think I deleted the message chunking by accident o.O

**Comment by AceLikesGhosts** [comment]
Current **idea** solution as towards us not receiving our own messages is to create an anonymous client which receives our message event, along with our normal authenticated one to send messages. That way, with our anonymous client we receive all messages and can counter act our authenticated client not receiving the onMessage event when we call ChatClient#say. Will start prototyping and committing most likely tomorrow.

**Comment by AceLikesGhosts** [comment]
Untested IMPL, although should work; will test when I'm home.

**Comment by AceLikesGhosts** [comment]
Merging master into this branch messed made both the checks fail, I hate you.

**Comment by AceLikesGhosts** [comment]
Still untested code, IMPL should work hopefully. (:

**Comment by iSlammedMyKindle** [comment]
> Merging master into this branch messed made both the checks fail, I hate you.

leOof

**Comment by iSlammedMyKindle** [comment]
There are a lot of changes here! What I'm gonna do is load them onto my local code editor since basically all the core files were changed

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:26]: As far as I know, utf-8 is the default, so it's why I left it out initially :+1:

**Comment by iSlammedMyKindle** [comment]
That's interesting... it looks like twurple still requires that we have our own oauth solution, but it does provide a way to login and refresh tokens... either they did that because they are expecting people to have their own solution, or that it was just too hard to make themselves XP

**Review by iSlammedMyKindle — CHANGES_REQUESTED** [review]
There's a lot of stuff that needs to be changed

I'm about half-way through, but instead of holding comments, I'll let you see what I have so far

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:26]: That's just how Twurple does it in their docs, that was basically ripped from there. (:

File: src/oauth.ts
Comment (AceLikesGhosts) [reply src/oauth.ts:56]: We have to change it to camel case from snake case or else Twurple will throw an error. It only supports camel case; also tried to throw in Hydra because that's just a quality lib. (not biased)

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:65]: This can officially be be a one line arrow function! eslint approves

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:15]: We 100% need to rename some of the functions and stuff; I made a function to wrap everything due to just having a import statements in our index.ts would just be.. weird looking compared to how it is now. Like.. you can't say the second one ain't look better ```ts
import './clients/discord';
import './clients/twitch';

// compared to

import registerTwitch from './clients/twitch';
import registerDiscord from './clients/discord';

registerTwitch();
registerDiscord();
```

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:38]: The way that auth is implemented right now is just how Twurple recommends/shows in their docs; Removed `DEV_TWITCH_TOKEN` due to in this we are already just storing our refresh token and other data by default, will add it back ._.

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:30]: We could make it an interface, but what's the point in an interface which we use maybe thrice in a codebase?

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:26]: Actually! Looking at it, we have to pass utf-8 or else FS thinks we're passing in a buffer instead of a path! I love Node.

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:39]: Reimplemented `DEV_TWITCH_TOKEN`; tested and it works.

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts]: Made it an interface after I realized that thrice is a lot. <3

File: src/oauth.ts
Comment (AceLikesGhosts) [reply src/oauth.ts]: Implemented [this](https://discord.com/channels/993857770233278474/998098626335219742/1016856237725466674) as you said; it works, on you if it has bad performance. <3

File: config.json
Comment (AceLikesGhosts) [reply config.json:13]: Yeah, merged master into this; gotta keep the new features in master in my PR!

**Comment by AceLikesGhosts** [comment]
> There's a lot of stuff that needs to be changed
> 
> I'm about half-way through, but instead of holding comments, I'll let you see what I have so far

Rereview \>:( <3

**Comment by AceLikesGhosts** [comment]
Oh yeah, by the way; tested and it works!

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [reply src/clients/twitch.ts:15]: That first one looks very nice!

**Review by iSlammedMyKindle — COMMENTED** [review]
I don't know why my comments aren't coming in immediately; I'll try to see if I can get that to work

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:15]: No.

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:102]: +ratio

File: src/clients/discord.ts
Comment (iSlammedMyKindle) [inline src/clients/discord.ts:113]: bruh

Comment (AceLikesGhosts) [reply src/clients/discord.ts:113]: That actually works! It didn't on the starting commits and I forgot to remove that comment (:

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:100]: I haven't loaded the code yet; but I don't know if we even need this anymore

If what I heard was true in that this library fixes our trailing ID problem, we could probably get rid of this statement alltogether

Comment (AceLikesGhosts) [reply src/clients/twitch.ts:100]: You right, I'm pretty sure we don't; Kept it due to not wanting to somehow break something, cause like the entire onMessage event handler is horrible to read ._.

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:1]: configFile might be the only one you need to destructure here, according to the search tool, `genericPromiseError` and `manageMsgCache` only come up with two results, so they could use the original `bridge` import ^_^

Comment (AceLikesGhosts) [reply src/clients/twitch.ts:1]: The Bridge import is just the `bridge` object as defined in.. bridge.ts, which is an object which does not contain `genericPromiseError`, or `manageMsgCache`. The `bridge` import is just the default import from bridge!
We could add `genericPromiseError` or `manageMsgCache` to the `bridge` default export, although that'd be pretty yikes.

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:69]: This is actually one of the things that was used to tackle the trailing ID bug; if it doesn't exist here, wer're good to remove this!

Comment (AceLikesGhosts) [reply src/clients/twitch.ts:69]: Tried not to remove anything which was not necessary to remove; if I'm recalling correctly that would not even send to begin with. Will remove. 

Also "we're"*

File: src/oauth.ts
Comment (iSlammedMyKindle) [inline src/oauth.ts:51]: eyyyyyy - it's the thing o.o

Comment (AceLikesGhosts) [reply src/oauth.ts:51]: ur fault

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [reply src/clients/twitch.ts:1]: Ah gotcha XP

**Comment by iSlammedMyKindle** [comment]
Gonna give it a launch and perform a blackbox test!

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [reply src/clients/twitch.ts:39]: This is a very interesting implementation - nice work!

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:50]: If I'm understanding this correctly; are we writing to this regardless if `DEV_TWITCH_TOKEN` is false?
Best to keep our footprint low under environments that are read-only

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:61]: So it looks like there are two clients, one that has authentication or logged in as a user, and one that doesn't. Based on what I'm reading the anon client is taking in messages, while the auth is sending them

My question is how come we separate them in the first place? Doing this increases the workload at the very least memory wise.

An explanation above line 61 in the code itself would probably be useful here too, since this might be a little odd at first to read :mag:

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:127]: I think this is useful to confirm if we're grabbing the token; otherwise we'd be worried that deploying over 50 times exhausts our token supply

I don't mind how it's phrased, weather it be this or another, but having some indication that we're using a previously saved token is pretty nice

**Review by iSlammedMyKindle — CHANGES_REQUESTED** [review]
Looking great so far! I can't approve it yet because the bug where discord receives a twitch copy of a discord message still persists.

When that's fixed, I'll be sure to have another look

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:61]: If we listen for messages with an authed ChatClient we will not receive the event when we call ChatClient#say, so we have to create another ChatClient which is not authed/doesn't have the say method called on it to recieve the event.

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:127]: Impled.

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:29]: Thanks! Looks good

File: src/clients/twitch.ts
Comment (iSlammedMyKindle) [inline src/clients/twitch.ts:57]: The glorious land of jank solutions brings forth beauty <3

My co-workers don't like it when I do this, however I like it, so this gets a pass XD

Comment (AceLikesGhosts) [reply src/clients/twitch.ts:57]: look. if it works it works!

File: src/clients/twitch.ts
Comment (AceLikesGhosts) [reply src/clients/twitch.ts:29]: ez pz

**Comment by iSlammedMyKindle** [comment]
General tip - In a biz environment, commit descriptions can actually be very useful to determine how did what at a certain time. One of our favorite extensions at our company is something called "git lens", where each line you click on shows when it was last changed, who made it, and the description that it contained

At my work we also use a ticketing system (which I'm not interested in using here), that gives you an exact story/development task ID to lookup elsewhere, so you can then ultimately look at documentation on it's initial progress

So yeah, commit comments are really useful!

**Comment by AceLikesGhosts** [comment]
no

**Review by iSlammedMyKindle — APPROVED** [review]
My brain is wiped out, but it looks like everything's great!

Hopefully this will be the last giant change for a while, I think it's good to have smaller changes so multiple people don't have file conflicts

**Pull request merged**
